From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Mon, 12 Oct 2026 21:32:39 -0700
Subject: [PATCH] mshtml: keyboard, storage and mouse event members

KeyboardEvent.initKeyboardEvent keeps the key, location, repeat and locale it
is given and sets the modifiers from the list. StorageEvent.initStorageEvent
and storageArea work. A mouse event's offsetX/offsetY are the client position
relative to the target's padding edge (they were 0). charCode no longer
repeats keyCode. A message event's data of a non-string type reads as its
string form instead of E_NOTIMPL.

diff --git a/dlls/mshtml/htmlevent.c b/dlls/mshtml/htmlevent.c
index 43968d3..ccc46b7 100644
--- a/dlls/mshtml/htmlevent.c
+++ b/dlls/mshtml/htmlevent.c
@@ -508,6 +508,7 @@ typedef struct {
     BSTR old_value;
     BSTR new_value;
     BSTR url;
+    IHTMLStorage *storage_area;
 } DOMStorageEvent;
 
 static DOMStorageEvent *DOMStorageEvent_from_DOMEvent(DOMEvent *event)
@@ -3028,24 +3029,64 @@ static HRESULT WINAPI DOMMouseEvent_get_y(IDOMMouseEvent *iface, LONG *p)
     return IDOMMouseEvent_get_pageY(iface, p);
 }
 
+/* The position inside the target: the client position relative to the target's padding edge. */
+static HRESULT mouse_event_get_offset(DOMMouseEvent *This, BOOL vertical, LONG *p)
+{
+    IEventTarget *target = NULL;
+    IHTMLElement2 *elem2;
+    IHTMLElement *elem;
+    IHTMLRect *rect;
+    LONG client, origin = 0, border = 0;
+    HRESULT hres;
+
+    hres = vertical ? IDOMMouseEvent_get_clientY(&This->IDOMMouseEvent_iface, &client)
+                    : IDOMMouseEvent_get_clientX(&This->IDOMMouseEvent_iface, &client);
+    if(FAILED(hres))
+        return hres;
+
+    if(SUCCEEDED(IDOMEvent_get_target(&This->ui_event.event.IDOMEvent_iface, &target)) && target) {
+        if(SUCCEEDED(IEventTarget_QueryInterface(target, &IID_IHTMLElement, (void**)&elem))) {
+            if(SUCCEEDED(IHTMLElement_QueryInterface(elem, &IID_IHTMLElement2, (void**)&elem2))) {
+                if(SUCCEEDED(IHTMLElement2_getBoundingClientRect(elem2, &rect))) {
+                    if(vertical) {
+                        IHTMLRect_get_top(rect, &origin);
+                        IHTMLElement2_get_clientTop(elem2, &border);
+                    }else {
+                        IHTMLRect_get_left(rect, &origin);
+                        IHTMLElement2_get_clientLeft(elem2, &border);
+                    }
+                    IHTMLRect_Release(rect);
+                }
+                IHTMLElement2_Release(elem2);
+            }
+            IHTMLElement_Release(elem);
+        }
+        IEventTarget_Release(target);
+    }
+
+#ifdef SG_MUTANT_MSHTML_OFFSET_CLIENT
+    origin = border = 0;
+#endif
+    *p = client - origin - border;
+    return S_OK;
+}
+
 static HRESULT WINAPI DOMMouseEvent_get_offsetX(IDOMMouseEvent *iface, LONG *p)
 {
     DOMMouseEvent *This = impl_from_IDOMMouseEvent(iface);
 
-    FIXME("(%p)->(%p) returning 0\n", This, p);
+    TRACE("(%p)->(%p)\n", This, p);
 
-    *p = 0;
-    return S_OK;
+    return mouse_event_get_offset(This, FALSE, p);
 }
 
 static HRESULT WINAPI DOMMouseEvent_get_offsetY(IDOMMouseEvent *iface, LONG *p)
 {
     DOMMouseEvent *This = impl_from_IDOMMouseEvent(iface);
 
-    FIXME("(%p)->(%p) returning 0\n", This, p);
+    TRACE("(%p)->(%p)\n", This, p);
 
-    *p = 0;
-    return S_OK;
+    return mouse_event_get_offset(This, TRUE, p);
 }
 
 static HRESULT WINAPI DOMMouseEvent_get_pageX(IDOMMouseEvent *iface, LONG *p)
@@ -3197,6 +3238,13 @@ typedef struct {
     DOMUIEvent ui_event;
     IDOMKeyboardEvent IDOMKeyboardEvent_iface;
     nsIDOMKeyEvent *nsevent;
+
+    /* set by initKeyboardEvent: the engine's key events have no setters for these */
+    BOOL init_called;
+    BSTR key;
+    ULONG location;
+    VARIANT_BOOL repeat;
+    BSTR locale;
 } DOMKeyboardEvent;
 
 static inline DOMKeyboardEvent *impl_from_IDOMKeyboardEvent(IDOMKeyboardEvent *iface)
@@ -3215,6 +3263,10 @@ static HRESULT WINAPI DOMKeyboardEvent_get_key(IDOMKeyboardEvent *iface, BSTR *p
 
     TRACE("(%p)->(%p)\n", This, p);
 
+    if(This->init_called) {
+        *p = SysAllocStringLen(This->key, SysStringLen(This->key));
+        return *p ? S_OK : E_OUTOFMEMORY;
+    }
 
     nsAString_Init(&key_str, NULL);
     nsres = nsIDOMKeyEvent_GetKey(This->nsevent, &key_str);
@@ -3229,6 +3281,11 @@ static HRESULT WINAPI DOMKeyboardEvent_get_location(IDOMKeyboardEvent *iface, UL
 
     TRACE("(%p)->(%p)\n", This, p);
 
+    if(This->init_called) {
+        *p = This->location;
+        return S_OK;
+    }
+
     nsres = nsIDOMKeyEvent_GetLocation(This->nsevent, &r);
     if(NS_FAILED(nsres))
         return E_FAIL;
@@ -3309,6 +3366,11 @@ static HRESULT WINAPI DOMKeyboardEvent_get_repeat(IDOMKeyboardEvent *iface, VARI
 
     TRACE("(%p)->(%p)\n", This, p);
 
+    if(This->init_called) {
+        *p = This->repeat;
+        return S_OK;
+    }
+
     nsres = nsIDOMKeyEvent_GetRepeat(This->nsevent, &r);
     if(NS_FAILED(nsres))
         return E_FAIL;
@@ -3337,38 +3399,78 @@ static HRESULT WINAPI DOMKeyboardEvent_getModifierState(IDOMKeyboardEvent *iface
     return S_OK;
 }
 
+static BOOL modifiers_list_has(const WCHAR *list, const WCHAR *name)
+{
+    size_t len = wcslen(name);
+
+    while(list && *list) {
+        while(*list == ' ')
+            list++;
+        if(!wcsncmp(list, name, len) && (!list[len] || list[len] == ' '))
+            return TRUE;
+        while(*list && *list != ' ')
+            list++;
+    }
+    return FALSE;
+}
+
 static HRESULT WINAPI DOMKeyboardEvent_initKeyboardEvent(IDOMKeyboardEvent *iface, BSTR type,
         VARIANT_BOOL can_bubble, VARIANT_BOOL cancelable, IHTMLWindow2 *view, BSTR key,
         ULONG location, BSTR modifiers_list, VARIANT_BOOL repeat, BSTR locale)
 {
     DOMKeyboardEvent *This = impl_from_IDOMKeyboardEvent(iface);
-    BOOL ctrl = FALSE, alt = FALSE, shift = FALSE, meta = FALSE;
+    BSTR key_copy = NULL, locale_copy = NULL;
     nsAString type_str;
     nsresult nsres;
-    WCHAR *list, *tok, *ctx = NULL;
+    HRESULT hres;
 
     TRACE("(%p)->(%s %x %x %p %s %lu %s %x %s)\n", This, debugstr_w(type), can_bubble,
           cancelable, view, debugstr_w(key), location, debugstr_w(modifiers_list),
           repeat, debugstr_w(locale));
 
-    /* The modifiers are a space separated list of key names. */
-    if(modifiers_list && (list = wcsdup(modifiers_list))) {
-        for(tok = wcstok_s(list, L" ", &ctx); tok; tok = wcstok_s(NULL, L" ", &ctx)) {
-            if(!wcscmp(tok, L"Control")) ctrl = TRUE;
-            else if(!wcscmp(tok, L"Alt")) alt = TRUE;
-            else if(!wcscmp(tok, L"Shift")) shift = TRUE;
-            else if(!wcscmp(tok, L"Meta")) meta = TRUE;
-        }
-        free(list);
+    if(This->ui_event.event.target) {
+        TRACE("called on already dispatched event\n");
+        return S_OK;
+    }
+
+    if(view)
+        FIXME("view argument is not supported\n");
+
+    if(!(key_copy = SysAllocString(key ? key : L"")) || !(locale_copy = SysAllocString(locale ? locale : L""))) {
+        SysFreeString(key_copy);
+        return E_OUTOFMEMORY;
+    }
+
+    hres = IDOMEvent_initEvent(&This->ui_event.event.IDOMEvent_iface, type, can_bubble, cancelable);
+    if(FAILED(hres)) {
+        SysFreeString(key_copy);
+        SysFreeString(locale_copy);
+        return hres;
     }
 
-    /* The engine's initializer has no place for the location, repeat flag or locale; the view
-     * is not passed either (the mouse events do the same). */
     nsAString_InitDepend(&type_str, type);
-    nsres = nsIDOMKeyEvent_InitKeyEvent(This->nsevent, &type_str, !!can_bubble, !!cancelable, NULL,
-            ctrl, alt, shift, meta, 0, key && key[0] && !key[1] ? key[0] : 0);
+    nsres = nsIDOMKeyEvent_InitKeyEvent(This->nsevent, &type_str, !!can_bubble, !!cancelable,
+            NULL /* FIXME */, modifiers_list_has(modifiers_list, L"Control"),
+            modifiers_list_has(modifiers_list, L"Alt"), modifiers_list_has(modifiers_list, L"Shift"),
+            modifiers_list_has(modifiers_list, L"Meta"), 0, 0);
     nsAString_Finish(&type_str);
-    return NS_SUCCEEDED(nsres) ? S_OK : E_FAIL;
+    if(NS_FAILED(nsres)) {
+        FIXME("InitKeyEvent failed: %08lx\n", nsres);
+        SysFreeString(key_copy);
+        SysFreeString(locale_copy);
+        return E_FAIL;
+    }
+
+    SysFreeString(This->key);
+    SysFreeString(This->locale);
+    This->key = key_copy;
+    This->locale = locale_copy;
+    This->location = location;
+    This->repeat = repeat ? VARIANT_TRUE : VARIANT_FALSE;
+#ifndef SG_MUTANT_MSHTML_KEY_NOINIT
+    This->init_called = TRUE;
+#endif
+    return S_OK;
 }
 
 static HRESULT WINAPI DOMKeyboardEvent_get_keyCode(IDOMKeyboardEvent *iface, LONG *p)
@@ -3395,7 +3497,7 @@ static HRESULT WINAPI DOMKeyboardEvent_get_charCode(IDOMKeyboardEvent *iface, LO
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    nsres = nsIDOMKeyEvent_GetKeyCode(This->nsevent, &r);
+    nsres = nsIDOMKeyEvent_GetCharCode(This->nsevent, &r);
     if(NS_FAILED(nsres))
         return E_FAIL;
 
@@ -3453,7 +3555,8 @@ static HRESULT WINAPI DOMKeyboardEvent_get_locale(IDOMKeyboardEvent *iface, BSTR
     TRACE("(%p)->(%p)\n", This, p);
 
     /* No input locale is tracked for the engine's key events. */
-    *p = SysAllocString(L"");
+    *p = This->init_called && This->locale ? SysAllocStringLen(This->locale, SysStringLen(This->locale))
+                                           : SysAllocString(L"");
     return *p ? S_OK : E_OUTOFMEMORY;
 }
 
@@ -3672,8 +3775,18 @@ static HRESULT WINAPI DOMMessageEvent_get_data(IDOMMessageEvent *iface, BSTR *p)
     }
 
     if(V_VT(&This->data) != VT_BSTR) {
-        FIXME("non-string data\n");
-        return E_NOTIMPL;
+        /* the legacy accessor is a string: other values read as their string form */
+        VARIANT str;
+        HRESULT hres;
+
+        V_VT(&str) = VT_EMPTY;
+        hres = VariantChangeType(&str, &This->data, VARIANT_NOUSEROVERRIDE, VT_BSTR);
+        if(FAILED(hres)) {
+            WARN("data %s is not a string\n", debugstr_variant(&This->data));
+            return hres;
+        }
+        *p = V_BSTR(&str);
+        return S_OK;
     }
 
     return (*p = SysAllocString(V_BSTR(&This->data))) ? S_OK : E_OUTOFMEMORY;
@@ -4014,8 +4127,17 @@ static HRESULT WINAPI DOMStorageEvent_get_url(IDOMStorageEvent *iface, BSTR *p)
 static HRESULT WINAPI DOMStorageEvent_get_storageArea(IDOMStorageEvent *iface, IHTMLStorage **p)
 {
     DOMStorageEvent *This = impl_from_IDOMStorageEvent(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    if((*p = This->storage_area))
+        IHTMLStorage_AddRef(*p);
+    return S_OK;
+}
+
+static BSTR copy_event_string(BSTR str)
+{
+    return SysAllocStringLen(str, SysStringLen(str));
 }
 
 static HRESULT WINAPI DOMStorageEvent_initStorageEvent(IDOMStorageEvent *iface, BSTR type, VARIANT_BOOL can_bubble,
@@ -4023,9 +4145,45 @@ static HRESULT WINAPI DOMStorageEvent_initStorageEvent(IDOMStorageEvent *iface,
                                                        BSTR newValueArg, BSTR urlArg, IHTMLStorage *storageAreaArg)
 {
     DOMStorageEvent *This = impl_from_IDOMStorageEvent(iface);
-    FIXME("(%p)->(%s %x %x %s %s %s %s %p)\n", This, debugstr_w(type), can_bubble, cancelable,
+    BSTR key = NULL, old_value = NULL, new_value = NULL, url = NULL;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s %x %x %s %s %s %s %p)\n", This, debugstr_w(type), can_bubble, cancelable,
           debugstr_w(keyArg), debugstr_w(oldValueArg), debugstr_w(newValueArg), debugstr_w(urlArg), storageAreaArg);
-    return E_NOTIMPL;
+
+    if(This->event.target) {
+        TRACE("called on already dispatched event\n");
+        return S_OK;
+    }
+
+    if((keyArg && !(key = copy_event_string(keyArg))) || (oldValueArg && !(old_value = copy_event_string(oldValueArg)))
+       || (newValueArg && !(new_value = copy_event_string(newValueArg))) || (urlArg && !(url = copy_event_string(urlArg))))
+        hres = E_OUTOFMEMORY;
+    else
+        hres = IDOMEvent_initEvent(&This->event.IDOMEvent_iface, type, can_bubble, cancelable);
+    if(FAILED(hres)) {
+        SysFreeString(key);
+        SysFreeString(old_value);
+        SysFreeString(new_value);
+        SysFreeString(url);
+        return hres;
+    }
+
+    SysFreeString(This->key);
+    SysFreeString(This->old_value);
+    SysFreeString(This->new_value);
+    SysFreeString(This->url);
+    This->key = key;
+    This->old_value = old_value;
+    This->new_value = new_value;
+    This->url = url;
+
+    if(storageAreaArg)
+        IHTMLStorage_AddRef(storageAreaArg);
+    if(This->storage_area)
+        IHTMLStorage_Release(This->storage_area);
+    This->storage_area = storageAreaArg;
+    return S_OK;
 }
 
 static const IDOMStorageEventVtbl DOMStorageEventVtbl = {
@@ -4060,6 +4218,8 @@ static void DOMStorageEvent_destructor(DispatchEx *dispex)
     SysFreeString(storage_event->old_value);
     SysFreeString(storage_event->new_value);
     SysFreeString(storage_event->url);
+    if(storage_event->storage_area)
+        IHTMLStorage_Release(storage_event->storage_area);
     DOMEvent_destructor(dispex);
 }
 
@@ -4125,9 +4285,17 @@ dispex_static_data_t MouseEvent_dispex = {
     .iface_tids   = MouseEvent_iface_tids,
 };
 
+static void DOMKeyboardEvent_destructor(DispatchEx *dispex)
+{
+    DOMKeyboardEvent *This = DOMKeyboardEvent_from_DOMEvent(DOMEvent_from_DispatchEx(dispex));
+    SysFreeString(This->key);
+    SysFreeString(This->locale);
+    DOMEvent_destructor(dispex);
+}
+
 static const dispex_static_data_vtbl_t DOMKeyboardEvent_dispex_vtbl = {
     .query_interface  = DOMKeyboardEvent_query_interface,
-    .destructor       = DOMEvent_destructor,
+    .destructor       = DOMKeyboardEvent_destructor,
     .traverse         = DOMKeyboardEvent_traverse,
     .unlink           = DOMKeyboardEvent_unlink
 };
